server/request.go

```diff
@@ -208,41 +212,45 @@ func (r *RpcRequest) handle_sendRawTransaction() {
 		return
 	}
 
-	r.log("txHash: %s - from: %s / to: %s / nonce: %d / gasPrice: %s", r.tx.Hash(), r.txFrom, AddressPtrToStr(r.tx.To()), r.tx.Nonce(), BigIntPtrToStr(r.tx.GasPrice()))
+	r.log("txHash: %s - from: %s / to: %s / nonce: %d / gasPrice: %s", r.tx.Hash(), r.txFrom, utils.AddressPtrToStr(r.tx.To()), r.tx.Nonce(), utils.BigIntPtrToStr(r.tx.GasPrice()))
 	txFromLower := strings.ToLower(r.txFrom)
 
 	if r.tx.Nonce() >= 1e9 {
 		r.log("tx rejected - nonce too high: %d - %s from %s", r.tx.Nonce(), r.tx.Hash(), txFromLower)
-		delete(State.accountWithNonceFix, txFromLower)
 		r.writeRpcError("tx rejected - nonce too high")
+		err = RState.DelNonceFixForAccount(txFromLower)
+		if err != nil {
+			r.logError("redis:DelAccountWithNonceFix failed: %v", err)
+		}
 		return
 	}
 
 	// Remember time when tx was received
 	txHashLower := strings.ToLower(r.tx.Hash().Hex())
-	State.txHashToUser[txHashLower] = NewStringWithTime(txFromLower)
-	State.userLatestTxHash[txFromLower] = NewStringWithTime(txHashLower)
+	err = RState.SetSenderOfTxHash(txHashLower, txFromLower)
+	if err != nil {
+		r.logError("redis:SetSenderOfTxHash failed: %v", err)
+	}
 
 	if isOnOFACList(r.txFrom) {
 		r.log("BLOCKED TX FROM OFAC SANCTIONED ADDRESS")
 		r.writeHeaderStatus(http.StatusUnauthorized)
 		return
 	}
 
+	// Check if transaction needs protection
 	needsProtection := r.doesTxNeedFrontrunningProtection(r.tx)
 
-	// Special check for cancellation tx
-	if len(r.tx.Data()) == 0 && txFromLower == strings.ToLower(r.tx.To().Hex()) {
-		wasSentToRelay, found := State.userTxWithNonceSentToRelay[fmt.Sprintf("%s_%d", txFromLower, r.tx.Nonce())]
-		if found && wasSentToRelay.v {
-			// original tx was sent to relay
-			r.log("[cancel-tx] sending to relay for %s", txFromLower)
-			needsProtection = true
-		} else {
-			// original tx was sent to mempool, or not seen in rpc-endpoint
-			r.log("[cancel-tx] sending to mempool for %s", txFromLower)
-			needsProtection = false
+	// Check for cancellation-tx
+	if len(r.tx.Data()) <= 2 && txFromLower == strings.ToLower(r.tx.To().Hex()) {
+		requestDone := r.handleCancelTx() // returns true if tx was cancelled at the relay and response has been sent to the user
```

Comment:
What's the behaviour on the user's side here? Their cancellation tx looks successfully sent, right?

We could immediately fail that tx, although that might be confusing. I guess this returning a success and dropping them is best.

Reply:
The cancel-tx should succeed, and the initial tx (the one that was cancelled) should receive failed on next getTransactionReceipt